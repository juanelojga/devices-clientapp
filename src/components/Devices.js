import React from 'react';
import PropTypes from 'prop-types';

import Loading from './Loading';
import Device from './Device';

import styles from './Devices.module.scss';

const Devices = ({
  loading,
  devices,
  handleRemoveDevice,
  handleEditDevice
}) => (
  <div className={styles.devices}>
    {loading ? (
      <div className="u-center-text">
        <Loading loading={loading} />
      </div>
    ) : devices.length ? (
      devices.map(device => (
        <Device
          key={device.id}
          deviceId={device.id}
          systemName={device.system_name}
          type={device.type}
          hddCapacity={device.hdd_capacity}
          handleRemoveDevice={() => handleRemoveDevice(device)}
          handleEditDevice={() => handleEditDevice(device)}
        />
      ))
    ) : (
      <div className="u-center-text">
        <h2>No Results were found.</h2>
      </div>
    )}
  </div>
);

Devices.propTypes = {
  loading: PropTypes.bool,
  devices: PropTypes.array,
  handleRemoveDevice: PropTypes.func,
  handleEditDevice: PropTypes.func
};

export default Devices;
