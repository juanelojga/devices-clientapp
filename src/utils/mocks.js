export const DEVICES = [
  {
    id: '1',
    system_name: 'SUSAN-DESKTOP',
    type: 'WINDOWS_WORKSTATION',
    hdd_capacity: '128 GB'
  },
  {
    id: '2',
    system_name: 'MAC-LOCAL-FREDDY',
    type: 'MAC',
    hdd_capacity: '256 GB'
  },
  {
    id: '3',
    system_name: 'SMART-SERVER',
    type: 'WINDOWS_SERVER',
    hdd_capacity: '1024 GB'
  }
];
