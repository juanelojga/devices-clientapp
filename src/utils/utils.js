export const filterByType = (filter, values) =>
  values.filter(value => (filter === 'all' ? true : value.type === filter));

export const sortBy = (criteria, values) =>
  values.sort((a, b) => {
    const criteria1 = a[criteria].toUpperCase();
    const criteria2 = b[criteria].toUpperCase();
    return criteria1.localeCompare(criteria2, undefined, {
      numeric: true,
      sensitivity: 'base'
    });
  });
