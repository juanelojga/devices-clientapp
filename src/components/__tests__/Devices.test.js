import React from 'react';
import { render, fireEvent, cleanup, getByText } from '@testing-library/react';
import 'jest-dom/extend-expect';

import Devices from '../Devices';

import { DEVICES } from '../../utils/mocks';

afterEach(cleanup);

test('snapshot test', () => {
  const handleEditDevice = jest.fn();
  const handleRemoveDevice = jest.fn();
  const loading = false;

  const { container } = render(
    <Devices
      handleEditDevice={handleEditDevice}
      handleRemoveDevice={handleRemoveDevice}
      loading={loading}
      devices={DEVICES}
    />
  );
  expect(container).toMatchSnapshot();
});

test('snapshot test, empty devices', () => {
  const handleEditDevice = jest.fn();
  const handleRemoveDevice = jest.fn();
  const loading = false;

  const { container, getByText } = render(
    <Devices
      handleEditDevice={handleEditDevice}
      handleRemoveDevice={handleRemoveDevice}
      loading={loading}
      devices={[]}
    />
  );
  expect(getByText('No Results were found.')).toBeTruthy();
  expect(container).toMatchSnapshot();
});

test('snapshot test, loading', () => {
  const handleEditDevice = jest.fn();
  const handleRemoveDevice = jest.fn();
  const loading = true;

  const { container } = render(
    <Devices
      handleEditDevice={handleEditDevice}
      handleRemoveDevice={handleRemoveDevice}
      loading={loading}
      devices={[]}
    />
  );
  expect(container).toMatchSnapshot();
});

test('remove device', () => {
  const handleEditDevice = jest.fn();
  const handleRemoveDevice = jest.fn();
  const loading = false;

  const { getByTestId } = render(
    <Devices
      handleEditDevice={handleEditDevice}
      handleRemoveDevice={handleRemoveDevice}
      loading={loading}
      devices={DEVICES}
    />
  );
  fireEvent.click(getByTestId(`remove-${DEVICES[1].id}`));
  expect(handleRemoveDevice).toHaveBeenCalledTimes(1);
  expect(handleRemoveDevice).toHaveBeenCalledWith(DEVICES[1]);
});

test('edit device', () => {
  const handleEditDevice = jest.fn();
  const handleRemoveDevice = jest.fn();
  const loading = false;

  const { getByTestId } = render(
    <Devices
      handleEditDevice={handleEditDevice}
      handleRemoveDevice={handleRemoveDevice}
      loading={loading}
      devices={DEVICES}
    />
  );
  fireEvent.click(getByTestId(`edit-${DEVICES[0].id}`));
  expect(handleEditDevice).toHaveBeenCalledTimes(1);
  expect(handleEditDevice).toHaveBeenCalledWith(DEVICES[0]);
});
