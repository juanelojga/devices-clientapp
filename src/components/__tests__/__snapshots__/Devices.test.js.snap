// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div>
  <div
    class="devices"
  >
    <div
      class="row"
    >
      <div
        class="device"
      >
        <div
          class="text"
        >
          <h2
            class="heading"
          >
            SUSAN-DESKTOP
          </h2>
          <p>
            WINDOWS_WORKSTATION
          </p>
          <p>
            128 GB
          </p>
          <div
            class="actions"
          >
            <button
              class="btn btn-red"
              data-testid="remove-1"
            >
              Remove
            </button>
              
            <button
              class="btn btn-green"
              data-testid="edit-1"
            >
              Edit
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="device"
      >
        <div
          class="text"
        >
          <h2
            class="heading"
          >
            MAC-LOCAL-FREDDY
          </h2>
          <p>
            MAC
          </p>
          <p>
            256 GB
          </p>
          <div
            class="actions"
          >
            <button
              class="btn btn-red"
              data-testid="remove-2"
            >
              Remove
            </button>
              
            <button
              class="btn btn-green"
              data-testid="edit-2"
            >
              Edit
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="device"
      >
        <div
          class="text"
        >
          <h2
            class="heading"
          >
            SMART-SERVER
          </h2>
          <p>
            WINDOWS_SERVER
          </p>
          <p>
            1024 GB
          </p>
          <div
            class="actions"
          >
            <button
              class="btn btn-red"
              data-testid="remove-3"
            >
              Remove
            </button>
              
            <button
              class="btn btn-green"
              data-testid="edit-3"
            >
              Edit
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot test, empty devices 1`] = `
<div>
  <div
    class="devices"
  >
    <div
      class="u-center-text"
    >
      <h2>
        No Results were found.
      </h2>
    </div>
  </div>
</div>
`;

exports[`snapshot test, loading 1`] = `
<div>
  <div
    class="devices"
  >
    <div
      class="u-center-text"
    >
      <div
        class="css-0"
      >
        <div
          class="css-1w7hh59-style"
        />
        <div
          class="css-1i0ut9c-style"
        />
        <div
          class="css-1s2rl4q-style"
        />
      </div>
    </div>
  </div>
</div>
`;
